### ISSUE TITLE
outfit rule broken

### ISSUE THREAD

**Issue by kubax** [arkayenro/arkinventory#924]
outfit("") rules seems to be broken.

All my rules wich use outfit are disabled by arkinventory because of warnings from Blizzard_Deprecated addon.

[17:30:20] ArkInventory: WARNING> ...face\AddOns\Blizzard_Deprecated\Deprecated_7_2_0.lua:157: Usage: local equipmentSetID = C_EquipmentSet.GetEquipmentSetID(equipmentSetName)
[17:30:20] ArkInventory: WARNING> Regel 3 wird als defekt markiert und daher nicht länger ausgeführt, bis sie wieder korrigiert wurde

**Comment by Raejyn** [comment]
Same issue...

**Comment by arkayenro** [comment]
looks like blizzard rearranged things and may have broken their own code as thats a blizzard file, not one of mine.

i'll rewrite the rule code to use the new functions so it wont go through that code

**Comment by kubax** [comment]
i fixed it in Alpha r631

just add "if (items) then" bevore line 905 and add "end" after line 942.

So:
`		if (items) then
			for k, location in pairs( items ) do

				loc_id = nil
				bag_id = nil
				slot_id = nil
				id = nil

				player, bank, bags, void, slot, bag, voidtab, voidslot = EquipmentManager_UnpackLocation( location )

	--			if void then
	--				ArkInventory.Output( setname, ":", k, " -> [", player, ", ", bank, ", ", bags, ", ", void, "] [", bag, ".", slot, "] [", voidtab, ".", voidslot, "]" )
	--			end

				if void and voidtab and voidslot then
					loc_id = ArkInventory.Const.Location.Void
					bag_id = ArkInventory.Const.Offset.Void + voidtab
					slot_id = voidslot
					id = GetVoidItemInfo( voidtab, voidslot )
				elseif ( not bags ) and slot then
					loc_id = ArkInventory.Const.Location.Wearing
					bag_id = ArkInventory.Const.Offset.Wearing + 1
					slot_id = slot
					id = GetInventoryItemID( "player", slot )
				elseif bag and slot then
					loc_id, bag_id = ArkInventory.BlizzardBagIdToInternalId( bag )
					slot_id = slot
					id = GetContainerItemID( bag, slot )
				end

				if loc_id and bag_id and slot_id and id and ArkInventoryRules.Object.info.id and ArkInventoryRules.Object.loc_id == loc_id and ArkInventoryRules.Object.bag_id == bag_id and ArkInventoryRules.Object.slot_id == slot_id and id == ArkInventoryRules.Object.info.id then
					--ArkInventory.Output( setname, ":", k, " -> [", ArkInventoryRules.Object.h, " / ", id )
					tinsert( Outfits, string.trim( setname ) )
					--ArkInventory.Output( "found ", ArkInventoryRules.Object.h, " in set [", setname, ":", k, "] [", ArkInventoryRules.Object.loc_id, ".", ArkInventoryRules.Object.bag_id, ".", ArkInventoryRules.Object.slot_id, "]" )
					break
				end

			end
		end`

**Comment by arkayenro** [comment]
not sure how that would fix it as the error wasnt in that loop, and if items was nil then there would be nothing to loop through and it would skip that chunk of code

did you get some other error when using that alpha?

**Comment by kubax** [comment]
i tested with ArkInventory.Output(items) and it seems to be multiple arrays. first was completely emtpy, and last one was "nil" wich seems to throw the error. the code i tested was completely inspired (if not sayed stolen) from https://mods.curse.com/addons/wow/equipsee?comment=70 i don't have a clue what i am doing xD

Bugsack doesn#t show any errors from ArkInventory, so i go with "no" to the errors question :)

**Comment by arkayenro** [comment]
weird, items should be a flat array of 19 item location codes, it shouldnt be multiple arrays

**Comment by kubax** [comment]
it apears so to me.. as i said, i outputed items, and got this:

```
[17:24:44] ArkInventory: {  }
[17:24:44] ArkInventory: { [1]=1048577, [2]=1048578, [3]=1048579, [4]=1048580, [5]=1048581, [6]=1048582, [7]=1048583, [8]=1048584, [9]=1048585, [10]=1048586, [11]=1048587, [12]=1048588, [13]=1048589, [14]=1048590, [15]=1048591, [16]=1048592, [19]=1048595 }
[17:24:44] ArkInventory: { [1]=1048577, [2]=3146773, [3]=3146270, [4]=1048580, [5]=3145731, [6]=1048582, [7]=3146241, [8]=3146500, [9]=1048585, [10]=1048586, [11]=3145729, [12]=3146763, [13]=3145999, [14]=3146504, [15]=1048591, [16]=-1, [17]=3146497, [19]=1048595 }
[17:24:44] ArkInventory: { [1]=1048577, [2]=3145739, [3]=1048579, [4]=1048580, [5]=1048581, [6]=1048582, [7]=1048583, [8]=3146500, [9]=1048585, [10]=1048586, [11]=3145729, [12]=3146013, [13]=3146512, [14]=3145738, [15]=1048591, [16]=3145997, [19]=1048595 }
[17:24:44] ArkInventory: nil
```

Now that i think about it, it could be multiple calls of the Output command, as i am outputting it right after the definition of items, and thous must be one call for every Rule (i got 3 Rules) don't know where the empty one and the nil one are from.

But the function exits because of the nil one at the end. So this could be the point where this fix helps :)

**Comment by arkayenro** [comment]
yep, multiple calls, nested arrays would have nested { } brackets

i'll add the nil check to the code - not sure how you get that though, empty set maybe?

**Comment by kubax** [comment]
Oh, now everything is clear to me....

i got 5 set's but only 3 rules... 1 set was indeed empty (called nude.. what should i say... nude male dwarfs... raaawr....? xD)

One set seems to be broken (comes from dugi guides) i can't even delete it... this seeems to be the nil one.

**Comment by arkayenro** [comment]
should be a new alpha up if you want to check that one works

**Comment by Dragonspawn** [comment]
I had the same problem as Kubax. Alpha release has resolved the equipment rules for me :)

**Comment by kursion** [comment]
It seems that the `Rules` with `Outfit("...something...")` isn't working in the bank.

I've set a custom rule with called ILVL with `outfil("ilvl")`
I've apply this rule to group in the bag. It's working, all the item from outfit `ilvl` are going to that group.
But in the bank, only one item is going in the group.

Where is the algorithm in the code that handle this ? Thanks
